auth: Split SAML connector code to own type

camscale opened the pull request:
Create a new type - `SAMLAuthService` - and move the implementation of
the SAML connector from the auth Service to this new type. This
decoupling will make it easier to move out of tree in a later commit.

This also lays the basic groundwork for running without the SAML
connector at all, returning a NotImplemented error for all the SAML
logic if the SAMLAuthService is not plugged in.

A number of names in the `auth` package have been exported to facilitate
this. While technically not necessary for this PR as the new type lives
in this `auth` package, these changes are to make the move of the SAML
connector code to the enterprise repository as simple as possible and be
almost a cut and paste. This allows the changes here to be reviewed more
easily, rather than requiring PRs be compared across two repositories.

Issue: https://github.com/gravitational/teleport.e/issues/525

---

NOTE: This will be easier to review going commit by commit. Each commit
is reasonably small and self-contained.

https://github.com/gravitational/teleport.e/pull/597 and https://github.com/gravitational/teleport/pull/18285 are follow-on PRs that move the refactored SAML connector to the enterprise repo and remove it from the OSS repo.

https://github.com/gravitational/teleport.e/pull/606 and https://github.com/gravitational/teleport/pull/18419 are follow-on PRs that move the SAML auth web handlers to the enterprise repo and remove it from the OSS repo.

camscale committed:
auth: Emit correct event for SAML connectors

Fix `UpsertSAMLConnector` and `DeleteSAMLConnector` to emit
`apievents.SAMLConnector*` events of `apievents.OIDCConnector*` events.
This bug seem to have no impact as the audit log still reads correctly,
probably due to the correct event type strings being used in the event
message. Both the SAML and OIDC event messages have the same structure,
so any serialised events will have been unmarshaled correctly.

Update test cases for all connector types to ensure the right event
messages are emitted.

camscale committed:
auth: Export ssoDiagContext for external use

Export the struct `ssoDiagContext` (now `SSODiagContext`) and its fields
and methods and decouple it from `auth.Server`. This is a helper type
for recording SSO test diagnostics to simplify passing SSO context
between the various SSO methods. It contains very little functionality
in its own right and does not expose much if any of the auth.Server
implementation by exporting it.

The decoupling is achieved by defining a small interface (one function)
for what is required by `SSODiagContext` to record the diagnostic info
in the backend. The `auth.Server` implements this interface already, so
it is a simple matter of updating the `NewSSODiagContext()` call sites
to pass the `auth.Server` instead of use it as a received. A couple of
test cases that wanted to stub out the backend now do so via the adapter
func `SSODiagServiceFunc`, implemented in the same vein as
`http.Handler/http.HandlerFunc`.

This will be needed to move some SSO plugins out of the OSS repo.

camscale committed:
auth: Export web session methods

Export `createWebSession` and `createSessionCert` for use by external
SSO providers. `createWebSession` has been renamed to
`CreateWebSessionFromReq` as there is already a `CreateWebSession`
method. Update `CreateWebSession` to use `CreateWebSessionFromReq`,
removing duplicated code.

camscale committed:
auth: Export unstable field of auth.Server

Export the `unstable` field of `auth.Server` so it can be accessed by
the SAML connector code when it is moved out-of-tree.

Further refactoring is possible to remove the unstable service
altogether, since it contains only one service -
`AssertionReplayService` and that is used only by the SAML connector.
However there is also an API endpoint that calls the service. That is
slated to be removed in v12, so the further refactoring can wait until
that is addressed.

camscale committed:
auth: Split SAML connector code to own type

Create a new type - `SAMLAuthService` - and move the implementation of
the SAML connector from the auth Service to this new type. This
decoupling will make it easier to move out of tree in a later commit.

This also lays the basic groundwork for running without the SAML
connector at all, returning a NotImplemented error for all the SAML
logic if the SAMLAuthService is not plugged in.

camscale committed:
auth: Use provided context instead of auth server context

Use the `ctx` parameter passed into `ValidateSAMLResponse` when calling
the event emitter instead of using the context stored in the auth
server. This helps decouple the SAML connector code from the auth
server, making it easier to move out of tree.

camscale committed:
auth: Export createUserParams struct and fields

Export the `createUserParams` struct and its fields so it can be used by
out-of-tree SSO connectors - soon to be SAML and OIDC. This is a
mechanical change with no logic changes. Names have been capitalised and
that's it.

camscale commented on lib/auth/sso_diag_context.go:
```diff
@@ -22,33 +22,48 @@
+		err := c.DiagService.CreateSSODiagnosticInfo(ctx, c.AuthKind, c.RequestID, c.Info)
 		if err != nil {
-			log.WithError(err).WithField("requestID", c.requestID).Warn("failed to write SSO diag info data")
+			log.WithError(err).WithField("requestID", c.RequestID).Warn("failed to write SSO diag info data")
```
This seems a little bogus. This code is providing structured data to logrus. It should be sanitizing it before outputting it and likely would for JSON output. If this does need to be sanitized, what sanitization do we need to perform?

zmb3 commented:
Haven't reviewed yet, but I wonder if we've gone so far to make a separate logical "service" for managing SAML connectors, maybe we should do the same at the gRPC level to start to improve the bloated auth.service.

camscale commented:
> Haven't reviewed yet, but I wonder if we've gone so far to make a separate logical "service" for managing SAML connectors, maybe we should do the same at the gRPC level to start to improve the bloated auth.service.

@zmb3 We may get there, but one step at a time. I would like to see the auth service split up - it's a bit of a beast to work with, but that is too much for this PR.

camscale committed:
auth: Export Server.withAuth

Export `Server.withAuth` so that it can be used by plugins in the
enterprise repository. This is already exported for proxy web plugins
and is now needed for the auth web plugins.

camscale committed:
auth: Export types for HTTP/JSON SAML interface

Export the types used in the HTTP/JSON SAML interface so that they can
be used by a plugin in the enterprise repository. Move them into
`saml.go` alongside the other SAML interface types.

camscale commented:
@tigrato @zmb3 Friendly ping

r0mant commented on lib/auth/auth.go:
```diff
@@ -474,6 +480,13 @@ type Server struct {
+// RegisterSAMLService registers ss as the SAMLService that provides the SAML
+// connector implementation. If a SAMLService has already been registered, this
+// will override the previous registration.
+func (a *Server) RegisterSAMLService(ss SAMLService) {
```
Nit: I think `SetSAMLService` would be more in line with how we usually name methods like this.

camscale replied:
Done: 22d628c80582a5f099e779a734062c5eb3575295

r0mant commented on lib/auth/saml.go:
```diff
@@ -69,11 +165,11 @@ func (a *Server) UpsertSAMLConnector(ctx context.Context, connector types.SAMLCo
 		return trace.Wrap(err)
 	}
-	if err := a.emitter.EmitAuditEvent(ctx, &apievents.OIDCConnectorDelete{
+	if err := sas.emitter.EmitAuditEvent(ctx, &apievents.SAMLConnectorDelete{
```
Nice typo catch.

camscale committed:
auth: Rename RegisterSAMLService to SetSAMLService

Addressing comment by @r0mant

Link: https://github.com/gravitational/teleport/pull/18255#discussion_r1023418452
